shelley/chain-and-ledger/executable-spec/test/Generator/Utxo.hs

```diff
@@ -0,0 +1,143 @@
+-- Also generates valid certificates.
+genTx :: LedgerEnv ShortHash MockDSIGN
+      -> (UTxOState, DPState)
+      -> KeyPairs
+      -> Gen Tx
+genTx (LedgerEnv _ slot _ pparams) (UTxOState utxo _ _ _, dpState) keys = do
+  keys' <- Gen.shuffle keys
+
+  -- inputs
+  (inputs, spendingBalance, spendWits) <- pickSpendingInputs keys' utxo
+
+  -- output addresses
+  recipientAddrs <- genRecipients keys'
+
+  -- certificates
+  (certs, certWits, deposits_) <- genDCerts keys' pparams dpState
+
+  -- attempt to make provision for certificate deposits (otherwise discard this generator)
+  when (spendingBalance < deposits_) Gen.discard
+  let balance_ = spendingBalance - deposits_
+
+  -- calc. fees and output amounts
+  let (fee, outputs) = calcFeeAndOutputs balance_ recipientAddrs
+
+  -- witnessed transaction
+  txBody <- genTxBody inputs outputs certs fee slot
+  let !wits = makeWitnessesVKey txBody (spendWits ++ certWits)
+      multiSig = Map.empty -- TODO @uroboros Generate multi-sig transactions
+
+  return (Tx txBody wits multiSig)
+
+-- | Generate a transaction body with the given inputs/outputs and certificates
+genTxBody
+  :: Set TxIn
+  -> [TxOut]
+  -> Seq DCert
+  -> Coin
+  -> Slot
+  -> Gen TxBody
+genTxBody inputs outputs certs fee slot = do
+  ttl <- genNatural 1 100
+  return $ TxBody
+             inputs
+             outputs
+             certs
+             Map.empty -- TODO @uroboros generate withdrawals
+             fee
+             (slot + Slot ttl)
+             emptyUpdate -- TODO @uroboros generate updates
+
+-- | Calculate the fee and distribute the remainder of the balance
+-- to the given addresses (as transaction outputs)
+calcFeeAndOutputs
+  :: Coin
+  -> [Addr]
+  -> (Coin, [TxOut])
+calcFeeAndOutputs balance_ addrs =
+  (fee, outputs)
+  where
+    -- TODO @uroboros fee=0 works for now since PParams minFeeA/B == 0.
+    --      We need to instead add a "draft" TxBody as argument, which will
+    --      give a measure of the minimum fee for this transaction.
+    --      This estimated fee can then be used to create the actual TxBody,
+    --      with the new fee baked in.
+    fee = 0
+    amountPerOutput = balance_ -- TODO @@@ splitCoin
+    outputs = ((`TxOut` amountPerOutput) <$> addrs)
+
+-- | Select unspent output(s) to serve as inputs for a new transaction
+--
+-- Returns the inputs along with the total spendable balance and spending keys.
+
+-- NOTE: this function needs access to the keys that the
+--       given utxo originated from (in order to produce the appropriate witnesses
+--       to spend these outputs)
+pickSpendingInputs
+  :: KeyPairs
+  -> UTxO
+  -> Gen (Set TxIn, Coin, [KeyPair])
+pickSpendingInputs keys (UTxO utxo) = do
+  -- TODO @uroboros select _multiple_ utxo outputs to serve as inputs!
+  (TxOut addr _) <- Gen.element (Map.elems utxo)
+
+  let --(select all inputs to the output addr
+      spendUtxo = UTxO $ Map.filter (\(TxOut a _) -> a == addr) utxo
```

Comment:
Can't you get the corresponding output from the second element you discarded above?

Reply:
yes, I've gone with your suggested approach. Before, I was following the pre-STS approach which was selecting an address and then filtering for (possibly multiple) inputs corresponding to that address.